src/engine/ContextDependentPiecewiseLinearConstraint.h

```diff
@@ -0,0 +1,246 @@
+       implemented.
+     */
+    virtual List<PhaseStatus> getAllCases() const = 0;
+
+    /*
+     * Returns case split corresponding to the given phase/id
+       TODO: Update the signature in PiecewiseLinearConstraint, once the new
+       search is integrated.
+     */
+    virtual PiecewiseLinearCaseSplit getCaseSplit( PhaseStatus caseId ) const = 0;
+
+
+    /*
+      Check if the constraint's phase has been fixed.
+    */
+    virtual bool phaseFixed() const = 0;
+
+    /*
+      If the constraint's phase has been fixed, get the (valid) case split.
+    */
+    virtual PiecewiseLinearCaseSplit getImpliedCaseSplit() const = 0;
+
+    /*
+      Register a bound manager. If a bound manager is registered,
+      this piecewise linear constraint will inform the tightener whenever
+      it discovers a tighter (entailed) bound.
+    */
+    void registerBoundManager( BoundManager *boundManager );
+
+    /*
+       Register context object. Necessary for lazy backtracking features - such
+       as _cdPhaseStatus and _activeStatus. Does not require initialization until
+       after pre-processing.
+     */
+    void initializeCDOs( CVC4::context::Context *context );
+
+    /*
+       Politely clean up allocated CDOs.
+     */
+    void cdoCleanup();
+
+    /*
+      Get the context object - debugging purposes only
+    */
+    CVC4::context::Context *getContext() const
+    {
+        return _context;
+    }
+
+    /*
+      Get the active status object - debugging purposes only
+    */
+    CVC4::context::CDO<bool> *getActiveStatusCDO() const
+    {
+        return _cdConstraintActive;
+    };
+
+    /*
+      Get the current phase status object - debugging purposes only
+    */
+    CVC4::context::CDO<PhaseStatus> *getPhaseStatusCDO() const
+    {
+        return _cdPhaseStatus;
+    }
+
+    /*
+      Get the infeasible cases object - debugging purposes only
+    */
+    CVC4::context::CDList<PhaseStatus> *getInfeasibleCasesCDList() const
+    {
+        return _cdInfeasibleCases;
+    }
+
+    /*
+       Mark that an exploredCase is infeasible, reducing the remaining search space.
+     */
+    void markInfeasible( PhaseStatus exploredCase );
+
+    /*
+      Returns phaseStatus representing next feasible case.
+      Returns PHASE_NOT_FIXED if no feasible case exists.
+      Assumes the contraint phase is PHASE_NOT_FIXED.
+      Worst case complexity O(n^2)
+      This method should be overloaded for MAX and DISJUNCTION constraints.
+     */
+    PhaseStatus nextFeasibleCase();
+
+    /*
+       Returns number of cases not yet marked as infeasible.
+     */
+    unsigned numFeasibleCases()
+    {
+        return _numCases - _cdInfeasibleCases->size();
+    }
+
+    /*
+        Returns true if there are feasible cases remaining.
+     */
+    bool isFeasible()
+    {
+        return numFeasibleCases() > 0u;
+    }
+
+    /*
+       Returns true if there is only one feasible case remaining.
+     */
+    bool isImplication()
```

Comment:
Make function const?
Also, we usually do == 1 ad not 1 ==.

Reply:
Fixed, changed style. The idea is to avoid accidental assignment, but I understand that with our style and reviews this should be unneccessary.